src/dir_contents.ml

```diff
@@ -434,23 +447,189 @@ let build_mlds_map (d : Super_context.Dir_with_jbuild.t) ~files =
       Some (doc, List.map (String.Map.values mlds) ~f:(Path.relative dir))
     | _ -> None)
 
-let get =
-  let cache = Hashtbl.create 32 in
-  fun sctx ~dir ->
-    Hashtbl.find_or_add cache dir ~f:(fun dir ->
-      match Super_context.stanzas_in sctx ~dir with
-      | None ->
-        { text_files = String.Set.empty
-        ; modules = lazy
-            { libraries = String.Map.empty
-            ; executables = String.Map.empty
-            ; rev_map = Module.Name.Map.empty
-            }
+module Dir_status = struct
+  type t =
+    | Empty_standalone of File_tree.Dir.t option
+    | Is_component_of_a_group_but_not_the_root of
+        Super_context.Dir_with_jbuild.t option
+    | Standalone of File_tree.Dir.t
+                    * Super_context.Dir_with_jbuild.t
+    | Group_root of File_tree.Dir.t
+                    * Super_context.Dir_with_jbuild.t
+
+  let is_standalone = function
+    | Standalone _ | Empty_standalone _ -> true
+    | _ -> false
+
+  let cache = Hashtbl.create 32
+
+  let analyze_stanzas stanzas =
+    List.fold_left stanzas ~init:(None, false) ~f:(fun acc stanza ->
+      let is_group_root, has_modules_consumers = acc in
+      match stanza with
+      | Include_subdirs (loc, x) ->
+        if Option.is_some is_group_root then
+          Loc.fail loc "The 'include_subdirs' stanza cannot appear \
+                        more than once";
+        (Some x, has_modules_consumers)
+      | Library _ | Executables _ | Tests _ ->
+        (is_group_root, true)
+      | _ -> acc)
+
+  let rec get sctx ~dir =
+    match Hashtbl.find cache dir with
+    | Some t -> t
+    | None ->
+      let t =
+        match
+          Option.bind (Path.drop_build_context dir)
+            ~f:(File_tree.find_dir (Super_context.file_tree sctx))
+        with
+        | None -> Empty_standalone None
```

Comment:
Is it intentional that this case doesn't populate the cache?

Reply:
It does, line 520